### PR TITLE
GH-42

### PR DESCRIPTION
**Pull request by CalebMaftei**
### Summary

*This Pull Request adds new functionality* that allows a team admin or system admin to adjust a dynamic team-welcome message. Previously, the only way a team message could be set was to adjust the config.json, and then redeploy the instance of mattermost. This has caused certain team admins over at Platform 1 to seek out a resolution that allows for the dynamic ability to set their team welcome message without adjusting the config.json.

*With this new functionality*, we have added 3 new slash commands that allow a user with the appropriate credentials to set, get, and delete what we are calling `dynamic team messages` with the following commands:

* `get_team_welcome` - Grabs the stored dynamic team message (if it exists), or retrieves the config.json message set. 
* `set_team_welcome [new message]` - Stores internally a new KV pair where the key is the team id joined with the prefix: "teammsg_", and where the value is the message. This does not override the config.json value though.
* `delete_team_welcome` - This deletes the KV pair set for the team without ever touching the config.json message (if it exists).

If a dynamic team message is found, then the get call will retrieve that first. If a user ever wanted to resort back to the config.json message, they will just need to delete the dynamic message. All values will defer to the config.json message if no dynamic message is found.

This new functionality also allows appropriate team admins to have access to the above slash commands, rather than permitting only system admins. 

Also, on line 115, I fixed a misspelling of the word 'welcome.' 

Please let me know if this is the direction you would like to pursue regarding the solution to issue 42! If there is anything that I left out of this PR, I would be more than happy to add that necessary information! Thank you! 

### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-welcomebot/issues/42

### CODE REVIEW

**Comment by CalebMaftei** [comment]
Issue #42 looks to be completed! Disregard this pull request.

**Comment by hanzei** [comment]
@CalebMaftei I don't see a PR that addresses #42. Could you please point me to that one?

File: server/command.go [hunk 1]
```diff
@@ -79,40 +122,160 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
...
+		return false
+	}
+	if !isSysadmin && !isTeamAdmin {
+		p.postCommandResponse(args, "You do not have the proper privileges to control this Team's welcome messages.")
+		return false
+	}
+	return true
+}
+
+func (p *Plugin) showDynamicMessages(args *model.CommandArgs) bool {
```
Comment (hanzei) [inline server/command.go:176]: Returning a `bool` make the code hard to read. Should it return a `(string, bool)` instead?

File: server/command.go [hunk 2]
```diff
@@ -79,40 +122,160 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
...
+	team, err := p.API.GetTeamByName(teamName)
+	if err != nil {
+		p.postCommandResponse(args, teamRetrievalErr, err)
+		return
+	}
+	teamID := team.Id
+
+	validPrivileges := p.validatePreviewPrivileges(teamID, args)
+	if !validPrivileges {
+		return
```
Comment (hanzei) [inline server/command.go:223]: No error for the user?

File: server/command.go [hunk 3]
```diff
@@ -121,6 +284,31 @@ func (p *Plugin) executeCommandList(args *model.CommandArgs) {
...
+			}
+			str.WriteString(fmt.Sprintf("\n * %s", team.Name))
+		}
+		page++
+		keys, err = p.API.KVList(page, 200)
+		if err != nil {
+			p.postCommandResponse(args, "Issue grabbing messages for teams.")
+			return
+		}
+	}
```
Comment (hanzei) [inline server/command.go:311]: This can be simplified by breaking the loop if 
```
if len(keys) < 200 {
    break
}
```

File: server/command.go [hunk 4]
```diff
@@ -121,6 +284,31 @@ func (p *Plugin) executeCommandList(args *model.CommandArgs) {
...
+		return
+	}
+	for len(keys) != 0 {
+		// retrieve id inside of kv pair
+		for _, key := range keys {
+			id := strings.ReplaceAll(key, welcomebotTeamWelcomeKey, "")
+			team, getTeamErr := p.API.GetTeam(id)
+			if getTeamErr != nil {
+				continue // the key is not corresponding to a team
+			}
```
Comment (hanzei) [inline server/command.go:302]: Can we check for the `welcomebotTeamWelcomeKey` prefix to ensure to only check these?

File: server/hooks.go [hunk 5]
```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
...
+		)
 	}
+
+	postChannel := &model.Post{
+		UserId:    p.botUserID,
+		ChannelId: teamMember.TeamId,
+		Message:   string(teamMessage),
+	}
+	time.Sleep(1 * time.Second)
+	_ = p.API.SendEphemeralPost(teamMember.UserId, postChannel)
```
Comment (hanzei) [inline server/hooks.go:63]: There is a lot of duplicate code here. Can this be avoided?

File: server/hooks.go [hunk 6]
```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
...
+	if appErr != nil {
+		fmt.Print("CHANNEL NAME: " + channelInfo.Name)
 		mlog.Error(
 			"error occurred while checking the type of the chanel",
 			mlog.String("channelId", channelMember.ChannelId),
 			mlog.Err(appErr),
 		)
 		return
-	} else if channelInfo.Type == model.CHANNEL_PRIVATE {
-		return
```
Comment (hanzei) [inline server/hooks.go:39]: Good catch, but let's move this out as unrelated change. See https://github.com/mattermost/mattermost-plugin-welcomebot/pull/79#issuecomment-881391007.

File: server/hooks.go [hunk 7]
```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
...
+				go p.processWelcomeMessage(*data, *message)
+			}
 		}
+		return
+	}
+
+	// We send a DM and an opportunistic ephemeral message to the channel. See
+	// the discussion at the link below for more details:
+	// https://github.com/mattermost/mattermost-plugin-welcomebot/pull/31#issuecomment-611691023
+	teamAddMessage := "# Welcome to the " + data.Team.DisplayName + " team!! \n\n"
```
Comment (hanzei) [inline server/hooks.go:44]: Why is this default message needed?

File: server/hooks.go [hunk 8]
```diff
@@ -68,10 +107,12 @@ func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.Ch
 	// We send a DM and an opportunistic ephemeral message to the channel. See
 	// the discussion at the link below for more details:
 	// https://github.com/mattermost/mattermost-plugin-welcomebot/pull/31#issuecomment-611691023
+	channelAddMessage := "# Welcome to the " + channelInfo.DisplayName + " channel. \n\n"
```
Comment (hanzei) [inline server/hooks.go:110]: See above. Why is this behavior change added here?

File: server/hooks.go [hunk 9]
```diff
@@ -85,6 +126,12 @@ func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.Ch
 		ChannelId: channelMember.ChannelId,
 		Message:   string(data),
 	}
-	time.Sleep(1 * time.Second)
+
+	if actor.Id == channelMember.UserId {
+		time.Sleep(3 * time.Second)
+	} else {
+		time.Sleep(15 * time.Second)
+	}
```
Comment (hanzei) [inline server/hooks.go:134]: What is the reasoning behind this change?

**Review by hanzei — CHANGES_REQUESTED** [review]
That is great PR. The code looks quite good :+1: 

Thanks also for adding this more granular permission checks :+1: 

One piece of feedback: It's idiomatic in go to use `err` as the variable name returned for errors. There are surely places where it's not needed, but you should stick with that guideline more.

File: README.md [hunk 10]
```diff
@@ -198,6 +198,19 @@ type MessageTemplate struct {
...
+## How to Dynamically Assign Team Messages
+
+This plugin also allows system admins and team admins to change the team message of an appropriate team through a slash command, rather than adjusting the config.json. 
+
+We have provided three options that are formatted identically to the channel welcomes:
+- `/welcomebot get_team_welcome` - prints the current team's welcome message if either the dynamic message exists, or the config.json message exists.
+- `/welcomebot set_team_welcome [welcome message]` - sets the current team's welcome message to the one defined in the slash command.
+- `/welcomebot delete_team_welcome` - deletes the current team's welcome message. This however does NOT delete any messages set inside of the config.json.
+
+For saving these messages, they are stored as KV pairs. Team message pairs have the prefix `teammsg_<team_id>` to distinguish them from the `chanmsg_<channel_id>` pairs.
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply README.md:210]: I added this in part to explain on a high level the functionality we added, as well as explain to team admins how they can use the new slash commands. However, we can easily remove that.

File: server/command.go [hunk 11]
```diff
@@ -23,6 +26,13 @@ const (
 	commandTriggerGetChannelWelcome    = "get_channel_welcome"
 	commandTriggerDeleteChannelWelcome = "delete_channel_welcome"
 	commandTriggerHelp                 = "help"
+	commandTriggerSetTeamWelcome       = "set_team_welcome"
+	commandTriggerGetTeamWelcome       = "get_team_welcome"
+	commandTriggerDeleteTeamWelcome    = "delete_team_welcome"
+
+	// Error Message Constants
+	teamRetrievalErr  = "error occurred while retrieving the welcome message for the team: `%s`"
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:34]: We added error message constants to reduce the number of string literals for scans such as fortify. However, since I think there is only one place that `teamRetrievalErr` is called, we could remove it from the constants section.

File: server/command.go [hunk 12]
```diff
@@ -57,6 +67,39 @@ func (p *Plugin) hasSysadminRole(userID string) (bool, error) {
 	return true, nil
 }
 
+func (p *Plugin) hasTeamAdminRole(userID string, teamID string) (bool, error) {
+	teamMember, appErr := p.API.GetTeamMember(teamID, userID)
+	if appErr != nil {
+		return false, appErr
+	}
+	if !strings.Contains(teamMember.Roles, "team_admin") {
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:75]: There most definitely is! I will get `"team_admin"` swapped out with `PERMISSIONS_TEAM_ADMIN` instead.

File: server/command.go [hunk 13]
```diff
@@ -209,13 +512,8 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
 
-	isSysadmin, err := p.hasSysadminRole(args.UserId)
-	if err != nil {
-		p.postCommandResponse(args, "authorization failed: %s", err)
-		return &model.CommandResponse{}, nil
-	}
-	if !isSysadmin {
-		p.postCommandResponse(args, "/welcomebot commands can only be executed by the user with system admin role")
+	errorOccurred := p.verifyUser(args)
+	if errorOccurred {
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:516]: I can swap that error with the  idiomatic`err` instead.

File: server/hooks.go (hunk 9 above)
Comment (CalebMaftei) [reply server/hooks.go:134]: This was added by another Dev I believe. However, I think it was due to a cluster related timing issue. I can look more into this and why it was added.

File: server/hooks.go (hunk 7 above)
Comment (CalebMaftei) [reply server/hooks.go:44]: This change I believe was added to give some kind of header to the welcome bot messages. However, if this ultimately feels like we are stripping freedoms to the welcome message, it can be thrown out.

File: server/hooks.go (hunk 8 above)
Comment (CalebMaftei) [reply server/hooks.go:110]: This change I believe was added to give some kind of header to the welcome bot messages. However, if this ultimately feels like we are stripping freedoms to the welcome message, it can be thrown out.

File: server/hooks.go (hunk 6 above)
Comment (CalebMaftei) [reply server/hooks.go:39]: Can definitely do that!

File: server/hooks.go (hunk 5 above)
Comment (CalebMaftei) [reply server/hooks.go:63]: I can go through this piece and refactor it to be less complicated!

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply README.md:210]: My suggestion would be to delete the paragraph about the KV pairs.

File: server/command.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/command.go:34]: I'm 1/5 that the code is easier to debug if the error strings are just literals. This way you can grep them.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply README.md:210]: I have removed the KV paragraph!

File: server/command.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:34]: I have replaced all the constants with the literal!

File: server/command.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:516]: just changed the name to `err`

File: server/command.go [hunk 14]
```diff
@@ -188,6 +376,121 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+			}
+		}
+		// if KV do not have message, and Config.json does not have message, then there is no message. Display Error case.
+		p.postCommandResponse(args, unsetMessageError)
+		return
+	}
+	p.postCommandResponse(args, string(data))
+}
+
+func (p *Plugin) verifyUser(args *model.CommandArgs) bool {
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:461]: the new name of the function is `checkCommandPermission`

File: server/command.go [hunk 15]
```diff
@@ -188,6 +376,121 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+
+	isTownSquare, channelErr := p.checkIfTownSquare(args.ChannelId)
+	if channelErr != nil {
+		p.postCommandResponse(args, "Channel authorization failed: %s", channelAdminErr)
+		return true
+	}
+	if !isSysadmin && !isTeamAdmin && isChannelAdmin && isTownSquare {
+		p.postCommandResponse(args, "/welcomebot commands cannot be executed by a channel admin in Town Square")
+		return true
+	}
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:490]: I believe the reason we added this extra check was specifically to bar any channel admins from trying to be able to change the welcome message for Town Square. 

However, if this feels like a moot point, or counter-productive to the goal of the welcome bot, we can remove the logic for town-square restrictions. What would you prefer?

File: server/hooks.go [hunk 16]
```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
...
 
 // UserHasJoinedChannel is invoked after the membership has been committed to
 // the database. If actor is not nil, the user was invited to the channel by
 // the actor.
-func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, _ *model.User) {
-	if channelInfo, appErr := p.API.GetChannel(channelMember.ChannelId); appErr != nil {
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, actor *model.User) {
+	channelInfo, appErr := p.API.GetChannel(channelMember.ChannelId)
+	if appErr != nil {
+		fmt.Print("CHANNEL NAME: " + channelInfo.Name)
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/hooks.go:72]: Yep! I will remove that real quick.

File: server/command.go [hunk 17]
```diff
@@ -188,6 +376,121 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+
+	if appErr != nil {
+		p.postCommandResponse(args, teamRetrievalErr, appErr)
+		return
+	}
+
+	if data == nil {
+		p.postCommandResponse(args, unsetMessageError)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:415]: We fetched for it to check if the message was already set. If it wasn't set, we would notify the user. However, in hindsight, we could easily just delete as default even if there is nothing to delete.

File: server/command.go [hunk 18]
```diff
@@ -79,40 +122,160 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
...
+		}
+		p.postCommandResponse(args, "error occurred while getting the Team Admin Role `%s`: `%s`", teamID, teamMemberErr)
+		return false
+	}
+	doesUserHavePrivileges := p.checkForProperPrivileges(args, args.UserId, teamID)
+
+	return doesUserHavePrivileges
+}
+
+func (p *Plugin) checkForProperPrivileges(args *model.CommandArgs, userID string, teamID string) bool {
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:157]: It definitely can. I just changed it to `isSystemOrTeamAdmin` instead.

File: server/command.go (hunk 2 above)
Comment (CalebMaftei) [reply server/command.go:223]: The response to the user if they do not have preview privileges is found inside the function `validatePreviewPrivileges`. That is where we check what kind of permissions error it is before returning it back to the user.

File: server/command.go (hunk 3 above)
Comment (CalebMaftei) [reply server/command.go:311]: I ended up removing this logic all together by retrieving a map with all the values beforehand. The new logic just loops through the map and using the key in each pair to append to the list of teams with welcome messages.

File: server/command.go (hunk 4 above)
Comment (CalebMaftei) [reply server/command.go:302]: I ended up removing this logic all together by retrieving a map with all the values beforehand. The new logic just loops through the map and using the key in each pair to append to the list of teams with welcome messages.

File: server/hooks.go (hunk 6 above)
Comment (CalebMaftei) [reply server/hooks.go:39]: It has been placed back into the code.

Comment (CalebMaftei) [reply server/hooks.go:39]: It has been placed back into the code .

File: server/hooks.go (hunk 9 above)
Comment (CalebMaftei) [reply server/hooks.go:134]: So digging into this deeper, the reason this was added was to prevent a race condition between the posted ephemeral message as well as the actual welcome message. The reasoning from our Dev was: "there was some race condition involving the welcome message being sent in a specific case. I think in one instance of a user being added to a channel, the ephemeral message was sent before the user was added to the channel, so they could never see it."

However, I think this can easily be mitigated by removing the ephemeral message as this was the issue with redundancy found inside of the team welcome messages as well. What are your thoughts?

File: server/hooks.go (hunk 5 above)
Comment (CalebMaftei) [reply server/hooks.go:63]: I removed the ephemeral message. I believe we originally had it so that the user would immediately be greeted by the welcome message, however the DM with the welcome-bot is plenty sufficient.

File: server/hooks.go [hunk 19]
```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
 		return
 	}
 
-	for _, message := range p.getWelcomeMessages() {
-		if message.TeamName == data.Team.Name {
-			go p.processWelcomeMessage(*data, *message)
+	key := fmt.Sprintf("%s%s", welcomebotTeamWelcomeKey, teamMember.TeamId)
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/hooks.go:20]: Created the helper functions at the bottom of `server/configuration.go` right next to the helper function that retrieves the welcome messages from the config.json file.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jfrerich) [reply README.md:210]: Hi, @CalebMaftei, can you push this commit to the PR branch so that we can see the update?

**Review by jfrerich — CHANGES_REQUESTED** [review]
Thanks for this PR, @CalebMaftei! I have some small suggestions on top of the request from Ben.

File: server/command.go [hunk 20]
```diff
@@ -23,6 +27,12 @@ const (
 	commandTriggerGetChannelWelcome    = "get_channel_welcome"
 	commandTriggerDeleteChannelWelcome = "delete_channel_welcome"
 	commandTriggerHelp                 = "help"
+	commandTriggerSetTeamWelcome       = "set_team_welcome"
+	commandTriggerGetTeamWelcome       = "get_team_welcome"
+	commandTriggerDeleteTeamWelcome    = "delete_team_welcome"
+
+	// Error Message Constants
+	unsetMessageError = "welcome message has not been set yet"
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:35]: The next push for this PR will have this change :)

File: server/command.go [hunk 21]
```diff
@@ -79,46 +122,151 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
 		if len(parameters) > 0 {
 			return "`delete_channel_welcome` command does not accept any extra parameters"
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:123]: The next push for this PR will have this change :) 

I did the same for set and get channel welcomes as well.

File: server/hooks.go [hunk 22]
```diff
@@ -17,18 +16,49 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
...
+		mlog.Error(
+			"error occurred while retrieving the welcome message",
+			mlog.String("teamId", teamMember.TeamId),
+			mlog.Err(appErr),
+		)
+		return
+	}
+
+	if teamMessage == nil {
+		// No dynamic welcome message for the given team, so we check if one as been set in the config.json
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/hooks.go:30]: The next push for this PR will have this change :)

File: server/command.go [hunk 23]
```diff
@@ -79,46 +122,151 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
...
+		return false
+	}
+	if !isSysadmin && !isTeamAdmin {
+		p.postCommandResponse(args, "You do not have the proper privileges to control this Team's welcome messages.")
+		return false
+	}
+	return true
+}
+
+// This retrives a map of team Ids with their respective welcome message
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:177]: The next push to this PR will have this spelling change :)

File: server/command.go [hunk 24]
```diff
@@ -79,46 +122,151 @@ func (p *Plugin) validateCommand(action string, parameters []string) string {
...
+	return doesUserHavePrivileges
+}
+
+// checks if the user has System or Team Admin access to the given team
+func (p *Plugin) isSystemOrTeamAdmin(args *model.CommandArgs, userID string, teamID string) bool {
+	isSysadmin, sysAdminError := p.hasSysadminRole(userID)
+	isTeamAdmin, teamAdminError := p.hasTeamAdminRole(userID, teamID)
+
+	if sysAdminError != nil {
+		p.postCommandResponse(args, "error occurred while getting the System Admin Role `%s`: `%s`", teamID, sysAdminError)
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:163]: I removed the team message header, as well as removed the unnecessary teamID for the `hasSystemadminRole()` function. Those changes should be in with the next push to this PR :)

File: server/command.go [hunk 25]
```diff
@@ -188,6 +336,116 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+	message := strings.SplitN(args.Command, "set_team_welcome", 2)[1]
+	message = strings.TrimSpace(message)
+
+	key := makeTeamWelcomeMessageKey(args.TeamId)
+	if appErr := p.API.KVSet(key, []byte(message)); appErr != nil {
+		p.postCommandResponse(args, "error occurred while storing the welcome message for the team: `%s`", appErr)
+		return
+	}
+
+	p.postCommandResponse(args, "stored the welcome message:\n%s", message)
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:356]: This change as well as other places in code that didn't distinguish channel from team welcome messages has been updated!

File: server/command.go [hunk 26]
```diff
@@ -188,6 +336,116 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+		p.postCommandResponse(args, "error occurred while retrieving the welcome message for the team: `%s`", appErr)
+		return
+	}
+
+	if appErr := p.API.KVDelete(key); appErr != nil {
+		p.postCommandResponse(args, "error occurred while deleting the welcome message for the team: `%s`", appErr)
+		return
+	}
+
+	p.postCommandResponse(args, "welcome message has been deleted")
```
[the remark this answers is not in the document]
Comment (CalebMaftei) [reply server/command.go:378]: This change will be in the next push for the PR :)

File: server/command.go [hunk 27]
```diff
@@ -188,6 +336,116 @@ func (p *Plugin) executeCommandDeleteWelcome(args *model.CommandArgs) {
...
+			}
+		}
+		// if KV do not have message, and Config.json does not have message, then there is no message. Display Error case.
+		p.postCommandResponse(args, unsetMessageError)
+		return
+	}
+	p.postCommandResponse(args, string(data))
+}
+
+func (p *Plugin) checkCommandPermission(args *model.CommandArgs) bool {
```
Comment (hanzei) [inline server/command.go:416]: I'm inclined to change the signature to 
```suggestion
func (p *Plugin) checkCommandPermission(args *model.CommandArgs) (bool, err) {
```

as it's more idiomatic to return an error instead of handling it in the function directly.

File: server/command.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/command.go:445]: I'm leaning toward removing the town-square restriction. Let's start simple for now.

File: server/command.go (hunk 2 above)
Comment (hanzei) [reply server/command.go:212]: Similar as in a comment above I think the signature

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for updating the PR. https://github.com/mattermost/mattermost-plugin-welcomebot/pull/81/files#r673361233 seems unresolved to me.

File: server/command.go (hunk 27 above)
Comment (CalebMaftei) [reply server/command.go:416]: New PR Will contain the signature fix, as well as the commented out section for the town square check! 

I also fixed the signature of the `validatePreviewPriviledges`.

Those changes should be seen with the next push to this PR! :)

**Comment by CalebMaftei** [comment]
Sorry about that. I thought that got pushed up. I will commit and push that change now. Sorry about that. It is commented out currently. If you would like, I can remove it all together as well.

File: README.md [hunk 28]
```diff
@@ -198,6 +198,17 @@ type MessageTemplate struct {
...
+<!-- ## How to Dynamically Assign Team Messages
+
+This plugin also allows system admins and team admins to change the team message of an appropriate team through a slash command, rather than adjusting the config.json. 
+
+We have provided three options that are formatted identically to the channel welcomes:
+- `/welcomebot get_team_welcome` - prints the current team's welcome message if either the dynamic message exists, or the config.json message exists.
+- `/welcomebot set_team_welcome [welcome message]` - sets the current team's welcome message to the one defined in the slash command.
+- `/welcomebot delete_team_welcome` - deletes the current team's welcome message. This however does NOT delete any messages set inside of the config.json.
+
+For saving these messages, they are stored as KV pairs. Team message pairs have the prefix `teammsg_<team_id>` to distinguish them from the `chanmsg_<channel_id>` pairs. -->
```
Comment (hanzei) [inline README.md:210]:
```suggestion
## How to Dynamically Assign Team Messages

This plugin also allows system admins and team admins to change the team message of an appropriate team through a slash command, rather than adjusting the config.json. 

We have provided three options that are formatted identically to the channel welcomes:
- `/welcomebot get_team_welcome` - prints the current team's welcome message if either the dynamic message exists, or the config.json message exists.
- `/welcomebot set_team_welcome [welcome message]` - sets the current team's welcome message to the one defined in the slash command.
- `/welcomebot delete_team_welcome` - deletes the current team's welcome message. This however does NOT delete any messages set inside of the config.json.

```

File: server/hooks.go (hunk 8 above)
Comment (hanzei) [reply server/hooks.go:100]: Yes, please throw it out.

File: server/hooks.go [hunk 29]
```diff
@@ -80,11 +111,18 @@ func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.Ch
...
+	// 	Message:   string(data),
+	// }
+
+	// if actor.Id == channelMember.UserId {
+	// 	time.Sleep(3 * time.Second)
+	// } else {
+	// 	time.Sleep(15 * time.Second)
+	// }
+
+	// _ = p.API.SendEphemeralPost(channelMember.UserId, postChannel)
```
Comment (hanzei) [inline server/hooks.go:127]: Why does this PR touch this code? Is it related to the changes you are doing here?

**Review by hanzei — CHANGES_REQUESTED** [review]

**Comment by raghavaggarwal2308** [comment]
Closing because of #143
